### PR TITLE
engine-storage: fix errored template becomes active

### LINKED ISSUE #7342: Templates become Inactive after restarting
See screenshots below

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Template
~~~

##### CLOUDSTACK VERSION

~~~
4.18.0.0
~~~

##### STEPS TO REPRODUCE

~~~

~~~

1. register a template with invalid url

image

2. restart mgmt server
image

##### EXPECTED RESULTS

~~~

~~~

##### ACTUAL RESULTS

~~~

~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #7342

When MS is restarted, failed non-public, non-featured, user templates are set to Active as their redownload is not attempted. This PR fixes the case.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

File: engine/storage/image/src/test/java/org/apache/cloudstack/storage/image/TemplateServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import org.mockito.Mockito;
+import org.mockito.Spy;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import com.cloud.storage.DataStoreRole;
+import com.cloud.storage.Storage;
+import com.cloud.storage.VMTemplateVO;
+
+@RunWith(MockitoJUnitRunner.class)
+public class TemplateServiceImplTest {
```
Comment (rohityadavcloud) [inline engine/storage/image/src/test/java/org/apache/cloudstack/storage/image/TemplateServiceImplTest.java:37]: Excellent to see tests @shwstppr !

File: engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java [hunk 2]
```diff
@@ -286,6 +286,23 @@ public void handleSysTemplateDownload(HypervisorType hostHyper, Long dcId) {
         }
     }
 
+    protected boolean isSkipTemplateStoreDownload(VMTemplateVO template, Long zoneId) {
+        if (template.isPublicTemplate()) {
```
Comment (weizhouapache) [inline engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java:290]: @shwstppr 
if an error template is public or featured, will it become Active ?

Comment (shwstppr) [reply engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java:290]: @weizhouapache no.
Server won't skip download if any of the following condition is true:
- Template is public
- Template is featured
- Template belongs to system
- Template is marked for the zone but not present in any of the zone secondary store

Depending upon download status template status will be set later

File: engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java [hunk 3]
```diff
@@ -286,6 +286,23 @@ public void handleSysTemplateDownload(HypervisorType hostHyper, Long dcId) {
...
+            return false;
+        }
+        if (template.isFeatured()) {
+            return false;
+        }
+        if (TemplateType.SYSTEM.equals(template.getTemplateType())) {
+            return false;
+        }
+        if (zoneId != null &&  _vmTemplateStoreDao.findByTemplateZone(template.getId(), zoneId, DataStoreRole.Image) == null) {
+            s_logger.debug(String.format("Template %s is not present on any image store for the zone ID: %d, its download cannot be skipped", template.getUniqueName(), zoneId));
```
Comment (Pearl1594) [inline engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java:300]: what if it is a direct download template?

File: engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java (hunk 2 above)
Comment (weizhouapache) [reply engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java:290]: it seems I misunderstood the code :-D

is there a scenario that template download is skipped but it is expected behaviour ? direct download ?

File: engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java (hunk 3 above)
Comment (shwstppr) [reply engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java:300]: Direct-download templates are skipped (already) at line 496

File: engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java (hunk 2 above)
Comment (shwstppr) [reply engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateServiceImpl.java:290]: Direct-download templates are skipped (already) at line 496

**Comment by shwstppr** [comment]
@Pearl1594 @weizhouapache are you guys okay to proceed with this?

**Comment by weizhouapache** [comment]
> @Pearl1594 @weizhouapache are you guys okay to proceed with this?

@shwstppr 
I will verify it

**Review by weizhouapache — APPROVED** [review]
Verified OK.

the template has been changed the correct status.
image

There might be a old bug (not caused by this PR)
the template was registered on
image
but  on UI
image

**Review by weizhouapache — APPROVED** [review]
Verified OK.

the template has been changed the correct status.
image

There might be a old bug (not caused by this PR)
the template was registered on
image
but  on UI
image

**Comment by shwstppr** [comment]
@weizhouapache it is probably taking the created date from `template_store_ref`. Will check and address separately

**Comment by weizhouapache** [comment]
> @weizhouapache it is probably taking the created date from `template_store_ref`. Will check and address separately

great, thanks @shwstppr 

if @Pearl1594 is ok with this PR, we can merge it.

**Review by Pearl1594 — APPROVED** [review]

**Pull request merged**